arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,295 @@
 #include <soc.h>
 
 #include <arch/arm/cortex_m/mpu/arm_core_mpu_dev.h>
+#include <linker/linker-defs.h>
 
 #define LOG_LEVEL CONFIG_MPU_LOG_LEVEL
 #include <logging/log.h>
 LOG_MODULE_REGISTER(mpu);
 
-#if defined(CONFIG_MPU_STACK_GUARD)
 /*
- * @brief Configure MPU stack guard
- *
- * This function configures per thread stack guards reprogramming the MPU.
- * The functionality is meant to be used during context switch.
- *
- * @param thread thread info data structure.
+ * Maximum number of dynamic memory partitions that may be supplied to the MPU
+ * driver for programming during run-time. Note that the actual number of the
+ * available MPU regions for dynamic programming depends on the number of the
+ * static MPU regions currently being programmed, and the total number of HW-
+ * available MPU regions. This macro is only used internally in function
+ * _arch_configurre_dynamic_mpu_regions(), to reserve sufficient area for the
+ * array of dynamic regions passed to the underlying driver.
  */
-void configure_mpu_stack_guard(struct k_thread *thread)
-{
-	u32_t guard_size = MPU_GUARD_ALIGN_AND_SIZE;
 #if defined(CONFIG_USERSPACE)
-	u32_t guard_start = thread->arch.priv_stack_start ?
-			    (u32_t)thread->arch.priv_stack_start :
-			    (u32_t)thread->stack_obj;
+#if defined(CONFIG_MPU_STACK_GUARD)
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM \
+	(CONFIG_MAX_DOMAIN_PARTITIONS + 2)
 #else
-	u32_t guard_start = thread->stack_info.start;
-#endif
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM \
+	(CONFIG_MAX_DOMAIN_PARTITIONS + 1)
+#endif /* CONFIG_MPU_STACK_GUARD */
+#else /* CONFIG_USERSPACE */
+#if defined(CONFIG_MPU_STACK_GUARD)
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM 1
+#else
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM 0
+#endif /* CONFIG_MPU_STACK_GUARD */
+#endif /* CONFIG_USERSPACE */
 
-	arm_core_mpu_disable();
-	arm_core_mpu_configure(THREAD_STACK_GUARD_REGION, guard_start,
-			       guard_size);
-	arm_core_mpu_enable();
-}
-#endif
+/* Convenience macros to denote the start address and the size of the system
+ * memory area, where dynamic memory regions may be programmed at run-time.
+ */
+#if defined(CONFIG_APP_SHARED_MEM)
+#define _MPU_DYNAMIC_REGIONS_AREA_START ((u32_t)&_app_smem_start)
```

Comment:
If you want the entire bounds of system RAM, _image_ram_start / _image_ram_end ought to be sufficient instead of all these ifdefs

Reply:
@andrewboie , I do not want the entire SRAM bounds; I need the boundaries, within which dynamic regions may be defined. Dynamic regions may be defined on Kernel SRAM (e.g. thread stacks), or Application Shared memory (e.g. app partitions).***

Now that I look more about it, the code assumes that these sections are adjacent. I can add a comment on the ARM linker script _not_ to change that. (Will track this externally)

*** Until we remove Application Memory, dynamic (mem domain) partitions may also be defined there. I do take care of this in the code.

Reply:
(addressed)